demos/http/http_demo_s3_download_multithreaded/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,1022 @@
+ * all communication is encrypted. After which, an HTTP thread is started which
+ * uses HTTP Client library API to send requests it reads from the request
+ * queue, and writes the responses to the response queue. The main thread sends
+ * requests on the request queue, which are used to download
+ * the S3 file by sending multiple range requests. While it is doing this, the
+ * main thread also reads responses from the response queue and prints them
+ * until the entire file is recieved. If any request fails, an error
+ * code is returned.
+ *
+ * @note This example is multi-threaded and uses statically allocated memory.
+ *
+ */
+int main( int argc,
+          char ** argv )
+{
+    /* Status to the host OS indicating a successful demo or not. */
+    int32_t returnStatus = EXIT_SUCCESS;
+
+    ( void ) argc;
+    ( void ) argv;
+
+    for( ; ; )
+    {
+        /* HTTPS Client library return status. */
+        HTTPStatus_t httpStatus = HTTP_SUCCESS;
+
+        /* The location of the path within the pre-signed URL. */
+        const char * pPath = NULL;
+
+        /* The length of the path within the prr-signed URL. This variable is
+         * defined in order to store the length returned from parsing the URL,
+         * but it is unused. The path used for the requests in this demo needs
+         * all the query information following the location of the object, to
+         * the end of the S3 presigned URL. */
+        size_t pathLen = 0;
+        /* The length of the Request-URI within string S3_PRESIGNED_GET_URL */
+        size_t requestLen = 0;
+
+        /* The transport layer interface used by the HTTP Client library. */
+        TransportInterface_t transportInterface = { 0 };
+        /* The network context for the transport layer interface. */
+        NetworkContext_t networkContext = { 0 };
+
+        /* Queue for HTTP requests. Requests are written by the main thread,
+         * and serviced by the HTTP thread. */
+        mqd_t requestQueue = -1;
+
+        /* Queue for HTTP responses. Responses are written by the HTTP thread,
+         * and read by the main thread. */
+        mqd_t responseQueue = -1;
+
+        /* PID of HTTP thread. */
+        pid_t httpThread = -1;
+
+        LogInfo( ( "HTTP Client multi-threaded S3 download demo using pre-signed URL:\n%s", S3_PRESIGNED_GET_URL ) );
+
+        /**************************** Parse Signed URL. ******************************/
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            /* Retrieve the path location from S3_PRESIGNED_GET_URL. This
+             * function returns the length of the path without the query into
+             * pathLen. */
+            httpStatus = getUrlPath( S3_PRESIGNED_GET_URL,
+                                     S3_PRESIGNED_GET_URL_LENGTH,
+                                     &pPath,
+                                     &pathLen );
+            requestLen = strlen( pPath );
```

Comment:
Why not just use the pathLen instead?

Reply:
Maybe add the same comment about the pathLen being unused to above this line for these onlookers to not get confused.
```suggestion
           /* The path used for the requests in this demo needs
            * all the query information following the location of the object, to
            * the end of the S3 presigned URL. */
            requestLen = strlen( pPath );
```